session/manager.go

```diff
@@ -105,6 +128,29 @@ func (sm *Manager) Reserve() error {
 	return nil
 }
 
+// earthly-specific
+func (sm *Manager) recordSessionStart(sessionID string) {
+	sm.history[sessionID] = &History{Start: time.Now()}
+}
+
+// earthly-specific
+func (sm *Manager) recordSessionEnd(sessionID string) {
+	if h := sm.history[sessionID]; h != nil {
+		h.End = time.Now()
+	}
+	for id, history := range sm.history {
+		if time.Since(history.Start) > sm.historyDuration {
+			delete(sm.history, id)
+		}
+	}
+}
+
+// GetSessionHistory returns a map of session ID to History entries
+// earthly-specific
+func (sm *Manager) GetSessionHistory() map[string]*History {
+	return sm.history
```

Comment:
this doesn't appear to be threadsafe, since `Controller.SessionHistory` can call this without acquiring `sm.mu`.

It might be best to lock, make a copy of the history, unlock, return the copy.

Reply:
Good point, let me see what I can do. It is a bit awkward because of how I assumed the lock is handled outside the session history code in the other functions. Maybe I'll re-work that so it just does the locking itself.

Reply:
Refactored so that locking is always handled by the session history functions.